### DISCUSSION TITLE
Simplest WebSockets example "fails" with a 403

### DISCUSSION THREAD

**Discussion by dAnjou** [fastapi/fastapi#11340] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()

@app.websocket("/foo")
async def foo(websocket):
    print("FOO")
    await websocket.accept()
    for line in ["line"]:
        await websocket.send_text(line)
```

### Description

```shell
pip install fastapi==0.110.0 uvicorn==0.29.0 websockets==12.0
uvicorn main:app --log-level debug
```
```shell
python -m websockets ws://127.0.0.1:8000/foo
```
Uvicorn output:
```
INFO:     Started server process [22379]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
DEBUG:    = connection is CONNECTING
DEBUG:    < GET /foo HTTP/1.1
DEBUG:    < host: 127.0.0.1:8000
DEBUG:    < upgrade: websocket
DEBUG:    < connection: Upgrade
DEBUG:    < sec-websocket-key: pJNin2YFaLlhwD3jzinClQ==
DEBUG:    < sec-websocket-version: 13
DEBUG:    < sec-websocket-extensions: permessage-deflate; client_max_window_bits
DEBUG:    < user-agent: Python/3.12 websockets/12.0
INFO:     ('127.0.0.1', 53509) - "WebSocket /foo" 403
DEBUG:    > HTTP/1.1 403 Forbidden
DEBUG:    > Date: Sun, 24 Mar 2024 17:09:06 GMT
DEBUG:    > Content-Length: 0
DEBUG:    > Content-Type: text/plain
DEBUG:    > Connection: close
DEBUG:    > [body] (0 bytes)
INFO:     connection rejected (403 Forbidden)
DEBUG:    x closing TCP connection
DEBUG:    = connection is CLOSED
INFO:     connection closed
```
Client output:
```
Failed to connect to ws://127.0.0.1:8000/foo: server rejected WebSocket connection: HTTP 403.
```

### Operating System

macOS

### FastAPI Version

0.110.0

### Pydantic Version

2.6.4

### Python Version

Python 3.12.2

**Comment by Kludex** [comment, marked as answer]
Missing the type hint on `websocket`:

```py
from fastapi import FastAPI, WebSocket

app = FastAPI()


@app.websocket('/foo')
async def foo(websocket: WebSocket):
    print('FOO')
    await websocket.accept()
    for line in ['line']:
        await websocket.send_text(line)
```

**Reply by dAnjou** [reply]
Huh, it _relies_ on a type _hint_? That should be mentioned in the documentation, no?

**Reply by Trolldemorted** [reply]
wouldn't HTTP 400 be more fitting than HTTP 403?

**Reply by Kludex** [reply]
It's according to the ASGI specification. Also, nothing FastAPI can do, it's on the server side.

**Comment by andzsinszan** [comment]
+1
Is this documented anywhere? 
The [type hint intro](https://fastapi.tiangolo.com/python-types/) does not explain much about the semantics, rather focuses is on the syntax. Which - to be honest - is not supposed to be in the scope of FastAPI's documentation.
